### PR TITLE
By default platform compat analyzer should be disabled for TFM < 5.0

### PR DESCRIPTION
**Pull request by buyaa-n**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/4091

### CODE REVIEW

File: src/Utilities/Compiler/Options/EditorConfigOptionNames.cs [hunk 1]
```diff
@@ -28,6 +28,11 @@ internal static partial class EditorConfigOptionNames
         /// </summary>
         public const string ExcludeAsyncVoidMethods = "exclude_async_void_methods";
 
+        /// <summary>
+        /// Boolean option to enable platform commpatibility analyzer for lower version than 5.0.
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/EditorConfigOptionNames.cs:32]: Please add a section at the end of https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#supported-editorconfig-options for the new option.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs [hunk 2]
```diff
@@ -162,6 +169,56 @@ static ImmutableArray<IMethodSymbol> GetOperatingSystemGuardMethods(IMethodSymbo
...
+        {
+            version = null;
+            for (int i = 0; i < tfm.Length; i++)
+            {
+                if (char.IsDigit(tfm[i]))
+                {
+                    if (i > 0)
+                    {
+                        platform = tfm.Substring(0, i);
+                        version = tfm[i..];
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:209]: Will this handle all flavored platforms mentioned here: https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md#design?

Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:209]: Yes, except `Tizen, Unity` i guess they are not expected to be supported though. Luckily we don't need to care about the different platforms, only `net5` section is matter from `net5.0-XYZ` so we are only checking if everything `before the first number` is equal to `net` and the `first number` is >= 5

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:209]: I thought the range expression will give entire substring starting from index i till the end.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 3]
```diff
@@ -41,6 +41,89 @@ public void MethodWithArrayArgument()
...
 
+        public static IEnumerable<object[]> Create_DifferentTfms()
+        {
+            yield return new object[] { "build_property.TargetFramework = net472", false };
+            yield return new object[] { "build_property.TargetFramework = netcoreapp1.0", false };
+            yield return new object[] { "build_property.TargetFramework = dotnet", false };
+            yield return new object[] { "build_property.TargetFramework = uap10.0", false };
+            yield return new object[] { "build_property.TargetFramework = netstandard2.1", false };
+            yield return new object[] { "build_property.TargetFramework = net5", true };
+            yield return new object[] { "build_property.TargetFramework = net99", true };
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:52]: Can you add tests for TFM value “net5.0” and “net5.0-xxx”?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs (hunk 2 above)
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:209]: Can we simplify the check to:
1. .Length >= 4 AND
2. .StartsWith(“net”, OrdinalIgnoreCase) AND
3 int.TryParse(tfm[3], out int major) AND
4. major >= 5

Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:209]: Yes, we don't really need to parse entire TFM

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs (hunk 3 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:52]: Sure

File: docs/Analyzer Configuration.md [hunk 4]
```diff
@@ -741,3 +741,15 @@ Configurable Rules: [CA1711](https://docs.microsoft.com/visualstudio/code-qualit
 Option Values: List (separated by '|') of allowed suffixes
 
 Example: `dotnet_code_quality.CA1711.allowed_suffixes = Flag|Flags`
+
+### Enable platform compatibility analyzer for TFMs <= net5.0
+
+Option Name: `enable_platform_analyzer`
+
+Configurable Rules: [CA14016](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1416)
```
Comment (Youssef1313) [inline docs/Analyzer Configuration.md:749]:
```suggestion
Configurable Rules: [CA1416](https://docs.microsoft.com/visualstudio/code-quality/ca1416)
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs [hunk 5]
```diff
@@ -162,6 +169,28 @@ static ImmutableArray<IMethodSymbol> GetOperatingSystemGuardMethods(IMethodSymbo
                     (method.Parameters.Length == 0 || method.Name.EndsWith(OptionalSuffix, StringComparison.Ordinal));
         }
 
+        private static bool PlatformAnalysisAllowed(AnalyzerOptions options, Compilation compilation, CancellationToken token)
+        {
+            var tfmString = options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.TargetFramework, compilation, token);
+
+            if (tfmString != null &&
+                tfmString.Length >= 4 &&
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:177]:
```suggestion
            if (tfmString?.Length >= 4 &&
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs [hunk 6]
```diff
@@ -162,6 +169,28 @@ static ImmutableArray<IMethodSymbol> GetOperatingSystemGuardMethods(IMethodSymbo
...
+                tfmString.StartsWith(Net, StringComparison.OrdinalIgnoreCase) &&
+                int.TryParse(tfmString[3].ToString(), out var major) &&
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: I'll let @buyaa-n make a call on whether we want to handle that case. If we wanted to, you can possibly write this as:
```suggestion
                tfmString.Length >= 5 &&  ? int.TryParse(tfmString[3..4], out var major) : int.TryParse(tfmString[3].ToString(), out major)
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 7]
```diff
@@ -41,6 +41,91 @@ public void MethodWithArrayArgument()
...
+        public static IEnumerable<object[]> Create_DifferentTfms()
+        {
+            yield return new object[] { "build_property.TargetFramework = net472", false };
+            yield return new object[] { "build_property.TargetFramework = netcoreapp1.0", false };
+            yield return new object[] { "build_property.TargetFramework = dotnet", false };
+            yield return new object[] { "build_property.TargetFramework = uap10.0", false };
+            yield return new object[] { "build_property.TargetFramework = netstandard2.1", false };
+            yield return new object[] { "build_property.TargetFramework = net5", true };
+            yield return new object[] { "build_property.TargetFramework = net5.0-windows", true };
+            yield return new object[] { "build_property.TargetFramework = net5.0-ios14.0", true };
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:53]: One additional scenario:
```suggestion
            yield return new object[] { "build_property.TargetFramework = net5", true };
            yield return new object[] { "build_property.TargetFramework = net5.0", true };
            yield return new object[] { "build_property.TargetFramework = net5.0-windows", true };
            yield return new object[] { "build_property.TargetFramework = net5.0-ios14.0", true };
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs [hunk 8]
```diff
@@ -41,6 +41,91 @@ public void MethodWithArrayArgument()
...
+            yield return new object[] { "build_property.TargetFramework = net472\ndotnet_code_quality.enable_platform_analyzer=false", false };
+            yield return new object[] { "build_property.TargetFramework = netcoreapp1.0\ndotnet_code_quality.enable_platform_analyzer=true", true };
+            yield return new object[] { "build_property.TargetFramework = netcoreapp1.0\ndotnet_code_quality.CA1416.enable_platform_analyzer=false", false };
+            yield return new object[] { "build_property.TargetFramework = dotnet\ndotnet_code_quality.enable_platform_analyzer=true", true };
+            yield return new object[] { "build_property.TargetFramework = uap10.0\ndotnet_code_quality.enable_platform_analyzer=false", false };
+            yield return new object[] { "build_property.TargetFramework = netstandard2.1\ndotnet_code_quality.enable_platform_analyzer=true", true };
+            yield return new object[] { "build_property.TargetFramework = netstandard2.1\ndotnet_code_quality.enable_platform_analyzer=false", false };
+            yield return new object[] { "build_property.TargetFramework = net5\ndotnet_code_quality.enable_platform_analyzer=false", true };
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.cs:96]:
```suggestion
            yield return new object[] { "build_property.TargetFramework = net5\ndotnet_code_quality.enable_platform_analyzer=false", true };
            yield return new object[] { "build_property.TargetFramework = net5.0\ndotnet_code_quality.enable_platform_analyzer=false", true };
            yield return new object[] { "build_property.TargetFramework = net5.0-windows\ndotnet_code_quality.enable_platform_analyzer=false", true };
            yield return new object[] { "build_property.TargetFramework = net5.0-ios14.0\ndotnet_code_quality.enable_platform_analyzer=false", true };
```

File: docs/Analyzer Configuration.md [hunk 9]
```diff
@@ -741,3 +741,15 @@ Configurable Rules: [CA1711](https://docs.microsoft.com/visualstudio/code-qualit
 Option Values: List (separated by '|') of allowed suffixes
 
 Example: `dotnet_code_quality.CA1711.allowed_suffixes = Flag|Flags`
+
+### Enable platform compatibility analyzer for TFMs <= net5.0
+
+Option Name: `enable_platform_analyzer`
```
Comment (mavasani) [inline docs/Analyzer Configuration.md:747]: Given the option is only applicable to pre-.NET5 TFMs, maybe we can make it explicit in the option name so it is clear that the option is not applicable to net5 or later TFMs? If you make this change, please update all uses of `enable_platform_analyzer` in this PR.
```suggestion
Option Name: `enable_platform_analyzer_on_pre_net5`
```

File: src/Utilities/Compiler/Options/EditorConfigOptionNames.cs [hunk 10]
```diff
@@ -28,6 +28,11 @@ internal static partial class EditorConfigOptionNames
         /// </summary>
         public const string ExcludeAsyncVoidMethods = "exclude_async_void_methods";
 
+        /// <summary>
+        /// Boolean option to enable platform commpatibility analyzer for TFMs with lower version than net5.0 (https://docs.microsoft.com/visualstudio/code-quality/ca1416).
+        /// </summary>
+        public const string EnablePlatformAnalyzer = "enable_platform_analyzer";
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/EditorConfigOptionNames.cs:34]: Should likely also rename this field if change the option name.

**Review by mavasani — APPROVED** [review]
Minor suggestions, but LGTM

File: docs/Analyzer Configuration.md (hunk 4 above)
Comment (buyaa-n) [reply docs/Analyzer Configuration.md:749]: Good catch, @mavasani JFYI the path copied from the above rules, all are having `en-us`

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: Good point, i better to account at least 2 digits

[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: as for valid TFM there is a char expected after `net5` i will change the condition on row 177 into `tfmString.Length >= 5`

File: docs/Analyzer Configuration.md (hunk 4 above)
Comment (mavasani) [reply docs/Analyzer Configuration.md:749]: Thanks, I'll create a separate PR to bulk remove `/en-us` from this page.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: > there is a char expected after net5

I thought just `net5` is a valid TFM value, so you need to allow for length = 4?

[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: Yeah seems it is valid

[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: The problem is there is .Net framework versions that are having 2 or more digits but accounted as lower than 5.0. like `net11`. See https://docs.microsoft.com/en-us/dotnet/standard/frameworks. I think i will keep it as is for now: 1st number >= 5

image

[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: If net11 refers to .NET Framework 1.1, then I wonder what will happen when we get to the "unified" .NET 11? (still a lot of years though) 😄

[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: hopefully, that time `.NET Framework` versions will be completely deprecated/removed

File: docs/Analyzer Configuration.md (hunk 9 above)
Comment (buyaa-n) [reply docs/Analyzer Configuration.md:747]: Yeah i have considered naming something like that and thought it is a too long name for devs to memorize, but after looking naming for other options seems we try to give a more descriptive name than keeping it short, will update

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:179]: Sounds good @buyaa-n

**Pull request merged**